### ISSUE TITLE
Slow Markdown Highlighting

### ISSUE THREAD

**Issue by pascalkuthe** [helix-editor/helix#4139]
### Summary

Since the recent update to the markdown grammer by @David-Else, I have noticed extreme slowdown on markdown files. For example if I hold down backspace and then let go, helix keeps deleting for roughly 0.5seconds.
I have a very beefy PC so this should really not be happening and do not see issues for any other language.
To isolate problems I have disabled the LSP but the problem persists until I turn of syntax highlighting (`set-language text`) completly.

Looking at the flamegraphs (with LSP disabled) I noticed that a ton of time seems to be spend on parsing TS compared to other languages (rust in this example)

opening and deleting some text in a **md** file:

![image](https://user-images.githubusercontent.com/61850714/194524594-6d232fb8-5e48-4409-8b13-b1a0bfec608c.png)

opening and deleting some text in a **rust** file:

![image](https://user-images.githubusercontent.com/61850714/194524545-7edb45df-ec21-428f-a892-5bb8fceb0538.png)

### Reproduction Steps

Open any markdown file with syntax highlighting enabled and started deleting text

### Platform

Linux

### Terminal Emulator

alacritty

### Helix Version

22.08.1-165-g4d4be0e8

**Comment by David-Else** [comment]
With `helix 22.08.1 (4d4be0e8)` I am not noticing:
> For example if I hold down backspace and then let go, helix keeps deleting for roughly 0.5seconds.

I am led to believe that Markdown is a bit slow for now and the author intends to improve it in the future. Did you try the flamegraph on the version before the update?

I am also on Linux (AlmaLinux 9) using the latest Kitty terminal.

**Comment by pascalkuthe** [comment]
Ok I tested again before the update to the markdown grammar.
I must have misremembered because the same slowdown still appears at that commit.
I guess the markdown grammar is just slow :/

I am very surprised that you do not notice this.
I have a very beefy PC (ryzen 9 3950X + GTX 1080 + 64GB ram, fedora 36).
Its very noticeable when comparing between editing rust and when continuously working on a somewhat large markdown file (1k loc). I sit at a constant 10% CPU utilization (which is a lot considering my CPU), the second I close the file (or switch to a rust file for example) the utilization drops to1%.

Just an example which might better highlight my problem: When editing the helix changelog and I edit the following line:

``` markdown
- The `extend_line` command has been replaced with `extend_line_below` and a new `extend_line` command now exists ([#3046](https://github.com/helix-editor/helix/pull/3046))
```

I start holding backspace at the EOL and let got at the first `(` so the line should look like this:

``` markdown
- The `extend_line` command has been replaced with `extend_line_below` and a new `extend_line` command now exists ([#3046](
```

but instead helix keeps deleting:
``` markdown
- The `extend_line` command has been replaced with `extend_line_below` and a new `extend_line`
```

This seems pretty extreme to me in terms of lag.
Perhaps something else is going on? I completely re-downloaded/recompiled all TS grammars and still get the same problem.

**Comment by David-Else** [comment]
I tried your test with the [changelog](https://raw.githubusercontent.com/helix-editor/helix/master/CHANGELOG.md) and my results were even worse, it deleted the entire line! Perhaps @MDeiml could comment?

I also tried on some technical writing I was working on which is mostly just prose of 700 lines with many embedded code blocks. I am using the Gnome system monitor and it reads 0.04-0.13% for Kitty running Helix with the file being edited/marksman language server running on an ancient Intel Haswell i7. I found no lag editing that.

Could it be all the links in the changelog are causing the slowdown? There is something going on...

**Comment by pascalkuthe** [comment]
Okay its definitely the links, by simply removing all of them completely fixes the lag.
Now we atleast know whats causing it :D

**Comment by David-Else** [comment]
I tried the same test with Neovim 0.7.1 and the results were much better, just a slight lag. The thread use was also different:
![Screenshot from 2022-10-07 12-57-28](https://user-images.githubusercontent.com/12832280/194547960-2c25a7fc-03de-46a3-aa84-fe10762657c9.png)

**Comment by pascalkuthe** [comment]
Are you sure that this is really the same grammar?
I removed all links from the CHANGELOG.md file and compared the flamegraphs:

**with links/lag**
![image](https://user-images.githubusercontent.com/61850714/194551475-310acb5d-1dda-4577-bc92-a7526efed29a.png)

**without links/lag**
![image](https://user-images.githubusercontent.com/61850714/194551573-a359e789-3279-438a-9ea8-2ed41cb664a3.png)

To me this looks like helix simply spends a ton of time calling the TS parser during the deletions.
I can't really see how nvim would improve upon this, if the TS implementation is the same.

**Comment by pascalkuthe** [comment]
I think I found the cause:

The links are highlighted using injections and injections are currently not parsed incrementally.
The original PR (#3108) by @MDeiml that added the split markdown parser actually mentions precisely this improvement:

> It's related to this TODO:
> 
> https://github.com/helix-editor/helix/blob/681c0a91dc36f9db1d23bd31ac1b66414d2a6099/helix-core/src/syntax.rs#L724
> 
> Just going one step further and also avoid parsing layers that are after the edits

I am looking into addressing this but its not trivial.

**Comment by pascalkuthe** [comment]
This is fixed by #4146.
While it looked a bit daunting at first, it actually was a pretty easy fix.
I am a bit surprised you are not seeing the same problems with `nvim`.
From nvim-treesitter/nvim-treesitter#2916 and neovim/neovim#14762 it does sound as tough the same issue appears there.
I setup `nvim` locally and while the problem wasn't quite as bad as with helix the lag was still noticeable.
However with helix and #4146 I get no lag at all (so much better then nvim)

**Comment by MDeiml** [comment]
I'm sorry I didn't participate in the discussion. Just wanted to say thank you, I think this was the best solution. The markdown grammar is slow, and that probably can't be fixed, but making parsing fully incremental should make any tree sitter parser sufficiently fast. I'm guessing this also sped up a lot of other grammars 👍
